src/lib.rs

```diff
@@ -1,9 +1,49 @@
 pub mod crypto;
+pub mod fast;
 mod impls;
 pub mod prelude;
-mod sequence_number;
-mod stable_hash;
 pub mod utils;
+mod verification;
+use prelude::*;
 
-pub use crate::sequence_number::{SequenceNumber, SequenceNumberInt};
-pub use crate::stable_hash::{StableHash, StableHasher};
+/// Like Hasher, but consistent across:
+/// * builds (independent of rustc version or std implementation details)
+/// * platforms (eg: 32 bit & 64 bit, x68 and ARM)
+/// * processes (multiple runs of the same program)
+pub trait StableHasher {
+    type Out;
+    type Addr: FieldAddress;
+
+    fn new() -> Self;
+    fn write(&mut self, field_address: Self::Addr, bytes: &[u8]);
+    fn mixin(&mut self, other: &Self);
+    fn finish(&self) -> Self::Out;
+
+    type Bytes: AsRef<[u8]>;
+    fn to_bytes(&self) -> Self::Bytes;
+    fn from_bytes(bytes: Self::Bytes) -> Self;
+}
+
+/// Like Hash, but consistent across:
+/// * builds (independent of rustc version or std implementation details)
+/// * platforms (eg: 32 bit & 64 bit, x68 and ARM)
+/// * processes (multiple runs of the same program)
+pub trait StableHash {
+    fn stable_hash<H: StableHasher>(&self, field_address: H::Addr, state: &mut H);
+}
+
+pub trait FieldAddress: Sized {
```

Comment:
This is a really important trait, and it should have comments on how to use it properly, and what `root` and `child` are used for; I don't understand at all what `unordered` is for, and how it should be used. I could definitely not implement this trait in a coherent way without such docs.

Reply:
I had not really intended this to be re-implemented, but I see your point. (Commenting is lacking across the board, really). Added lengthy comment.

Reply:
Beautiful. This explanation helps readability a lot!